services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -436,14 +436,15 @@ public AccessTokenResponseBuilder refreshAccessToken(KeycloakSession session, Ur
         return responseBuilder;
     }
 
-    private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel realm, RefreshToken refreshToken,
+    private void validateTokenReuseForRefresh(RealmModel realm, RefreshToken refreshToken,
         TokenValidation validation) throws OAuthErrorException {
         if (realm.isRevokeRefreshToken()) {
             AuthenticatedClientSessionModel clientSession = validation.clientSessionCtx.getClientSession();
             try {
-                validateTokenReuse(session, realm, refreshToken, clientSession, true);
-                int currentCount = clientSession.getCurrentRefreshTokenUseCount();
-                clientSession.setCurrentRefreshTokenUseCount(currentCount + 1);
+                validateTokenReuse(realm, refreshToken, clientSession);
+                int currentCount = clientSession.getRefreshTokenUseCount(refreshToken.getId());
+                clientSession.setRefreshTokenUseCount(refreshToken.getId(), currentCount + 1);
```

Comment:
I have the same concerns as you about the tons of client notes that can be created, I'm not sure if saving something like max usage in the refresh token will help since you still need to count the number of uses for each token. As you mentioned, cleaning up notes with expired tokens can help a bit, but it's not a solution for offline tokens.

Maybe a solution could be to save something like the `tab_id` in the refresh token, this way you can collect refresh tokens per tab and only keep in the client notes the last one used for each tab. WDYT?

Reply:
+1 for using `tabId` to track tokens. Moreover, I think that we should make sure that just one refresh token is valid per tabId. So for instance when refreshToken from `tab1` is refreshed and new refresh token `rt2` is generated for this tab, then after successfully refresh with using `rt2`, the previous `rt1` should not be valid anymore.

Otherwise, it would be possible that if some attacker manage to steal refresh token `rt1` and manage to successfully refresh it and generate refresh token `rt2-evil` and at the same time, the same refresh token `rt1` is used by correct user to generate another refresh token `rt2-good`, then both refresh tokens `rt2-evil` and `rt2-good` would be valid and would be possible to refresh them both successfully many time without detection that attacker successfully managed to steal the refresh token. This is exactly what we want to prevent here as IMO the main point of "revoke refresh token" feature is to prevent the case when someone manage to steal refresh token... We should make sure that there is only single "branch" valid for every `tabId` and hence there is still need to track somehow the refresh token per tab-id in clientSession though...

This is also aligned with the tooltips in admin console UI on fields `revokeRefreshTokenHelp` and `refreshTokenMaxReuseHelp`, which both mention that token is immediatelly revoked after it is successfully used.

So IMO it can be good if new methods on `AuthenticatedClientSessionModel` consider `tabId` rather than `refreshTokenId` and also that there is still `String getCurrentRefreshToken(String tabId)` . Also it looks that we may need to add something like `tabId` claim on refreshToken to make sure that refresh is successful just with refresh token, which matches `current refresh token` for the tab or where timestamp is equal to `iat` (so we may need separate note for last timestamp for the tab as well as field `timestamp` on `AuthenticatedClientSessionModel` cannot be used IMO due refreshes from many tabs).

Also using `tabId` should handle maximum limit for notes automatically as there is limit for maximum count of "Tabs" already per authentication session, which is 300 by default ( `InfinispanAuthenticationSessionProviderFactory.DEFAULT_AUTH_SESSIONS_LIMIT` ).